### ISSUE TITLE
Linux VRF support

### ISSUE THREAD

**Issue by vvidic** [perfsonar/pscheduler#780]
This is to discuss if we want to add Linux VRF support. It should be quite similar to the current pscheduler-context-linuxnns. Some more info:

https://www.kernel.org/doc/Documentation/networking/vrf.txt

```
DESCRIPTION
       A VRF provides traffic isolation at layer 3 for routing, similar to how a VLAN is used to isolate traffic
       at layer 2. Fundamentally, a VRF is a separate routing table. Network devices are associated with a VRF by
       enslaving the device to the VRF. At that point network addresses assigned to the device are local to the
       VRF with host and connected routes moved to the table associated with the VRF.

ip vrf exec [ NAME ] cmd ... - Run cmd against the named VRF
```

**Comment by mfeit-internet2** [comment]
If you have a way to test this, this would be trivial to add.

**Comment by mfeit-internet2** [comment]
A first cut of this is in the `issue-780` branch if you want to try it out.  The new context is called `linuxvrf`.

I adjusted everything I could find in the `debian` subdirectory except `debian/tests/task`, which I'll leave to you.

Make changes in this branch, let me know when you're good with it and I'll pull it into `4.2.0`.

**Comment by vvidic** [comment]
On Thu, Feb 28, 2019 at 06:32:03PM +0000, Mark Feit wrote:
> A first cut of this is in the `issue-780` branch if you want to try it out.  The new context is called `linuxvrf`.
> 
> I adjusted everything I could find in the `debian` subdirectory except `debian/tests/task`, which I'll leave to you.
> 
> Make changes in this branch, let me know when you're good with it and I'll pull it into `4.2.0`.

Great, thanks.  I will try to setup a host with vrf to test this.

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by vvidic** [comment]
On Thu, Feb 28, 2019 at 05:16:00PM +0000, Mark Feit wrote:
> If you have a way to test this, this would be trivial to add.

Seems the ip vrf command is available only in Ubuntu 18 at the
moment so we can use that for testing. When released CentOS 8
and Debian 10 should also work.

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by mfeit-internet2** [comment]
My CentOS 7 boxes have `ip vrf` as well.  If you have a test case for it like the one in NNS, I'll try it out.

**Comment by vvidic** [comment]
On Fri, Mar 01, 2019 at 06:37:07AM -0800, Mark Feit wrote:
> My CentOS 7 boxes have `ip vrf` as well.  If you have a test case for it like the one in NNS, I'll try it out.

Ok, pushed some changes to get it working with Ubuntu 18.
You can try if the test script works for CentOS 7 too.

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by vvidic** [comment]
Change merged into 4.2.0 branch, closing.

**Comment by mfeit-internet2** [comment]
**Release Notes:**

Added the `linuxvrf` context for changing VRFs.  (Currently functional only on distributions running Linux kernel >= 4.0.)
